src/plugins/chat.js

```diff
@@ -25,61 +25,51 @@ export class ChatPlugin extends Plugin {
     this.botNameRegex = new RegExp(`^(?:(?:@?${bot.name}|${bot.alias})[,\\s:.-]*)(.+)`, 'i');
 
     if (this.postConstructor) {
-      this.postConstructor.forEach(([fn, args]) => fn.bind(this)(...args));
+      this.postConstructor.forEach(([fn, args]) => fn.call(this, ...args));
     }
 
-    bot.emitter.on('receive-message', message => {
+    bot.emitter.on('receive-message', m => {
       // v = [
       //   validation function or regex,
       //   the response function from the plugin,
       //   the name of the function (the class.property name)
       // ];
       // as pushed by the listen/respond functions below
-      this.respondFunctions.forEach(v => this.process(...v, message, true));
-      this.listenFunctions.forEach(v => this.process(...v, message));
-    });
-  }
-
-  async process (validation, response, fnName, message, isResponder=false) {
-    try {
-      if (isResponder) {
-        const text = this.validateBotName(message);
-        if (!text) { return; }
-
-        message = new TextMessage({ ...message, text, direct: true });
-      } else if (message.whisper) {
 
-        // if it's a listener on a whisper, remove the botname first so it still
-        // works
-        const text = this.validateBotName(message);
-        if (!text) { return; }
+      // strip out the bot name from the beginning before processing
+      const text = this.stripLeadingBotName(m.text);
 
-        message = new TextMessage({ ...message, text, direct: true });
+      // If it had the bot name at the beginning, process the respond functions
+      if (text) {
+        // create a new message with the stripped text and pass it in
+        const message = new TextMessage({ ...m, text });
+        this.respondFunctions.forEach(v => this.process(...v, message));
+      } else if (m.whisper) {
+        this.respondFunctions.forEach(v => this.process(...v, m));
```

Comment:
Since adapter.js recieveWhisper prepends bot name we should move `if (m.whisper) { ... }` up into the previous block and iterate through the listenFunctions instead.